### ISSUE TITLE
Password reset not working

### ISSUE THREAD

**Issue by bethlakshmi** [bethlakshmi/gbe-divio-djangocms-python2.7#339]
This isn't a problem with #329 - a change I just deployed, tested, and reverted.
Even with the previous code, password change via email link doesn't work.  It *did* work - I tested it and tweaked the URL to work property after Mike Bergman had problems with it.

But it used to.

My local works.
Staging has a different problem that is getting in the way of me testing (the email data is unreadable in the Admin, so I can't get to the link).

I may try to actually escalate the #329 update because it lets users change their password, so we can manually change passwords, and then they can reset them...  

I'm also escalating the issue to Divio support - both issues:
 1- password reset link is received in email, but when you click it, you get an unsuccessful message
 2- email data is not available in the staging site's admin.

I'll also keep probing myself, but I feel like crap today, and don't have a lot of great ideas.  This is a tough one but important.

**Comment by bethlakshmi** [comment]
Hmm... I'm wondering about upgrading... I just found this:

https://django-allauth.readthedocs.io/en/latest/overview.html#supported-flows


This is a little bit like using a tank to drive to the grocery store.  The problem at hand is the forgotten password flow is broken.

What this is doing is replacing the standard django auth system with a system that lets a user connect to many, many different social accounts OR use traditional password auth.

We used to have that, way back when, but at some point, my method was too custom and broke.  This looks a hell of a lot smarter than my version.

Will try prototyping and seeing if I can get it working.

**Comment by burlexpo** [comment]
@bethlakshmi As we are looking at opening applications in mid-September, I'm kicking this up a priority level.

**Comment by bethlakshmi** [comment]
AHAHAHAHAHA!

I FIGURED IT OUT!  NOTHING IS WRONG!

Believe it or not - I do come back to this a lot because it really irritates me.

Our password reset *is* working.  My test method was not working.

Reason:

1. the password reset link takes a few min to be sent out - it gets queued, and then the queuing gets to it in a few minutes.
2. I get impatient - I want to get the thing verified and be done.  So I log in again (I have my password, after all!) and grab the link from the yet-to-be-sent email.
3. I try the link, it fails.

The problem is step number 2.  The way django computes the random hash used to make the link it sends to the user's email is that it uses a bunch of values _including last login timestamp_.  When the link is clicked in the email, then django recomputes the hash, and verifies the validity of the link.

By relogging in, I changed the last login timestamp, thereby making my link invalid.  So the step 3 failure was utterly normal.

I made a test user, repeated the flow and it worked perfectly.

I was my own enemy on testing this on and it's been working all along.

I retested using my own username/password and waiting for the email to come in - and that works too.
